d2/src/main/java/com/linkedin/d2/xds/XdsClient.java

```diff
@@ -0,0 +1,311 @@
+  {
+    void onChanged(RdsUpdate update);
+  }
+
+  interface CdsResourceWatcher extends ResourceWatcher
+  {
+    void onChanged(CdsUpdate update);
+  }
+
+  interface EdsResourceWatcher extends ResourceWatcher
+  {
+    void onChanged(EdsUpdate update);
+  }
+
+  interface ResourceUpdate
+  {
+  }
+
+  static final class LdsUpdate implements ResourceUpdate
+  {
+    // Http level api listener configuration.
+    final HttpConnectionManager _httpConnectionManager;
+
+    LdsUpdate(HttpConnectionManager httpConnectionManager)
+    {
+      _httpConnectionManager = httpConnectionManager;
+    }
+
+    public HttpConnectionManager getHttpConnectionManager()
+    {
+      return _httpConnectionManager;
+    }
+
+    @Override
+    public String toString()
+    {
+      return "LdsUpdate{" + "_httpConnectionManager=" + _httpConnectionManager + '}';
+    }
+
+    @Override
+    public int hashCode()
+    {
+      return Objects.hash(_httpConnectionManager);
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+      if (this == o)
+      {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass())
+      {
+        return false;
+      }
+      LdsUpdate ldsUpdate = (LdsUpdate) o;
+      return Objects.equals(_httpConnectionManager, ldsUpdate._httpConnectionManager);
+    }
+  }
+
+  static final class RdsUpdate implements ResourceUpdate
+  {
+    // The list of virtual hosts that make up the route table.
+    final List<VirtualHost> _virtualHosts;
+    final Map<String, Value> _serviceProperties;
+
+    RdsUpdate(List<VirtualHost> virtualHosts, Map<String, Value> serviceProperties)
+    {
+      _virtualHosts = Collections.unmodifiableList(new ArrayList<>(virtualHosts));
+      _serviceProperties = serviceProperties;
+    }
+
+    public List<VirtualHost> getVirtualHosts()
+    {
+      return _virtualHosts;
+    }
+
+    public Map<String, Value> getServiceProperties()
+    {
+      return _serviceProperties;
+    }
+
+    @Override
+    public String toString()
+    {
+      return "RdsUpdate{" + "_virtualHosts=" + _virtualHosts + ", _serviceProperties=" + _serviceProperties + '}';
+    }
+
+    @Override
+    public int hashCode()
+    {
+      return Objects.hash(_virtualHosts, _serviceProperties);
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+      if (this == o)
+      {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass())
+      {
+        return false;
+      }
+      RdsUpdate rdsUpdate = (RdsUpdate) o;
+      return Objects.equals(_virtualHosts, rdsUpdate._virtualHosts) && Objects.equals(_serviceProperties,
+          rdsUpdate._serviceProperties);
+    }
+  }
+
+  /** xDS resource update for cluster-level configuration. */
+  static final class CdsUpdate implements ResourceUpdate
+  {
+    final String _clusterName;
+    final String _edsServiceName;
+    final Map<String, Value> _clusterProperties;
+
```

Comment:
Probably we also want the Xds Cluster object here as well, if we want to explore using xds LB algorithms in the future.